Index: test.py
===================================================================
diff --git a/test.py b/test.py
new file mode 100644
--- /dev/null	
+++ b/test.py	
@@ -0,0 +1,26 @@
+import plotly.graph_objs as go
+import pysam
+import seaborn as sns
+import umap
+import vcf
+from plotly.offline import init_notebook_mode, iplot
+# import cyvcf2
+# cyvcf2.__version__
+
+import pandas as pd
+import numpy as np
+
+from sklearn.preprocessing import OneHotEncoder, LabelEncoder
+from sklearn.decomposition import PCA
+from sklearn.manifold import TSNE
+from MulticoreTSNE import MulticoreTSNE as TSNE
+
+
+from cyvcf2 import VCF, Writer
+
+import warnings
+warnings.filterwarnings('ignore',category=DeprecationWarning)
+
+import matplotlib.pyplot as plt
+import seaborn as sns
+plt.style.use('ggplot')
